backend/server/src/services/matchService.ts

```diff
@@ -193,9 +198,85 @@ export class MatchService {
     if (player1Points >= MAXIMUM_POINTS) {
       match.winner = player1.id;
       match.endTimestamp = new Date();
+      await this.createPlayoffSchedule(match.id, player1.id);
     } else if (player2Points >= MAXIMUM_POINTS) {
       match.winner = player2.id;
       match.endTimestamp = new Date();
+      await this.createPlayoffSchedule(match.id, player2.id);
+    }
+  }
+
+  private async createPlayoffSchedule(
+    matchId: Types.ObjectId,
+    winnerId: Types.ObjectId
+  ): Promise<void> {
+    const tournament = await TournamentModel.findOne({
+      matchSchedule: matchId
+    }).exec();
+    if (tournament?.type !== TournamentType.Playoff) {
+      return;
+    }
+
+    const playedMatches = await MatchModel.find({
+      _id: { $in: tournament.matchSchedule }
+    }).exec();
+
+    const currentMatch = playedMatches.find(
+      (match) => match.id.toString() === matchId.toString()
+    );
+    if (currentMatch === null || currentMatch === undefined) {
+      throw new NotFoundError({
+        message: "Match not found in tournament schedule"
+      });
+    }
+    const currentRound = currentMatch.tournamentRound;
+    const nextRound = currentRound + 1;
+
+    const winners = playedMatches
+      .filter((match) => match.tournamentRound === currentRound && match.winner)
+      .map((match) => match.winner)
+      .filter((winner): winner is Types.ObjectId => winner != null);
+
+    // Find eligible winners who don't have a match in the next round
+    const eligibleWinners = winners.filter((winner) => {
+      if (!winner) {
+        return false;
+      }
+      return !playedMatches.some(
+        (match) =>
+          match.tournamentRound === nextRound &&
+          match.players.some(
+            (player) => player.id.toString() === winner.toString()
+          )
+      );
+    });
+
+    // Pair current winner with eligible winners for the next round
+    for (const pairWithWinnerId of eligibleWinners) {
+      if (pairWithWinnerId && !pairWithWinnerId.equals(winnerId)) {
+        // Create a new match.
+        const newMatch = {
+          players: [
+            { id: winnerId, points: [], color: "red" },
+            { id: pairWithWinnerId, points: [], color: "white" }
+          ],
+          type: "playoff",
+          admin: null,
+          elapsedTime: 0,
+          timerStartedTimestamp: null,
+          tournamentRound: nextRound
+        };
+
+        const matchDocuments = await MatchModel.create(newMatch);
+        tournament.matchSchedule.push(matchDocuments.id);
+      }
+    }
+
+    // Save the tournament if new matches were added
+    if (eligibleWinners.length > 0) {
+      await TournamentModel.findByIdAndUpdate(tournament._id, {
+        matchSchedule: tournament.matchSchedule
```

Comment:
Cant you just call `save` on the tournament instead of querying it again?

Reply:
I don't know why i did this. :(